open-craft/edx-platform
# [SE-2968] Adds Custom Extra Fields to Course Blocks API through LMS Settings - Backported to Ironwood

Adds custom extra fields to Course Blocks API by specifying field names and the block type in the `COURSE_BLOCKS_API_EXTRA_FIELDS` LMS Settings variable.

`COURSE_BLOCKS_API_EXTRA_FIELDS` is a list of tuples. Each tuple contains two values, respectively:
1. Block Type (values separated by commas (,) or `*`)
2. Requested Field Name (string)

This PR backports the changes made in edx/edx-platform#24586 to Ironwood.2

**JIRA tickets**: SE-2968

**Upstream PR**: edx/edx-platform#24586

**Sandbox URL**: https://extrafields-backport.opencraft.hosting/

**Testing instructions**:

- Read through the code
- Run necessary tests and quality tests
    - `paver test_system`
    - `paver run_quality`
- Use this endpoint to test out API
    1. Login to the instance with the default `staff` credentials
    1. Navigate to [this course blocks REST API endpoint](https://extrafields-backport.opencraft.hosting/api/courses/v1/blocks/?course_id=course-v1%3AedX%2BDemoX%2BDemo_Course&all_blocks=True&requested_fields=other_course_settings,course_visibility), which returns the details about the Demonstration Course.
    1. Make sure that the block of `type=course` has `other_course_settings` and `course_visibility` in the JSON response
    1. Navigate to [this course blocks REST API endpoint](https://extrafields-backport.opencraft.hosting/api/courses/v1/blocks/?course_id=course-v1%3AedX%2BDemoX%2BDemo_Course&all_blocks=True&requested_fields=other_course_settings), which returns the details about the Demonstration Course.
    1. Make sure that the block of `type=course` has `other_course_settings` and **does not have** `course_visibility` in the JSON response
    1. Navigate to [this other course blocks REST API endpoint](https://extrafields-backport.opencraft.hosting/api/courses/v1/blocks/?course_id=course-v1%3AedX%2BDemoX%2BDemo_Course&all_blocks=True&depth=all&requested_fields=other_course_settings,course_visibility), which returns all blocks inside the Demonstration Course.
    1. Make sure that only blocks of `type=course` contain `other_course_settings`.

**Reviewers**

**Settings**
```yaml
EDXAPP_FEATURES:
  ENABLE_OTHER_COURSE_SETTINGS: true

EDXAPP_LMS_ENV_EXTRA:
  COURSE_BLOCKS_API_EXTRA_FIELDS: [
    ['course', 'course_visibility'],
    ['course', 'other_course_settings']
  ]
```
